doctor/lib/ocr_utils.py

```diff
@@ -0,0 +1,345 @@
+                layout=True, keep_blank_chars=True, y_tolerance=5, y_density=25
+            )
+        )
+    else:
+        doc_text = page.extract_text(
+            layout=True, keep_blank_chars=True, y_tolerance=5, y_density=25
+        )
+    return doc_text
+
+
+def page_images(page: pdfplumber.pdf.Page) -> bool:
+    """Does the page have images of a certain size
+
+    Meant to exclude images that might be lines
+    """
+    for img in page.images:
+        if (
+            img.get("width") / page.width * img.get("height") / page.height
+            > 0.1
+        ) or img.get("width") * img.get("width") > 10:
+            return True
+    return False
+
+
+def page_annotations(page: pdfplumber.pdf.Page) -> bool:
+    """Does the page have annotations which could contain text
+
+    :param page: pdf plumber
+    """
+    if page.annots:
+        anno_types = [
+            str(annot.get("data").get("Subtype")) for annot in page.annots
+        ]
+        if "/'FreeText'" in anno_types or "/'Widget'" in anno_types:
+            return True
+    return False
+
+
+def find_average_char_width(block_data: pd.Series) -> int:
+    """Average character width for a block of text
+
+    :param block_data:
+    :return: average character width
+    """
+    fd = block_data[block_data.text.str.len() > 0]
+    return (fd.width / fd.text.str.len()).mean()
+
+
+def validate_ocr_text(row: pd.Series, img: Image) -> pd.Series:
+    """Validate the OCR results
+
+    Take a look at our OCR and remove the bad results
+    and if possible reprocess words one by one
+
+    :param row: row of ocr results
+    :param img: Page image
+    :return: Updated row if necessary
+    """
+    if row["left"] < 370 and row["conf"] <= 40:
+        row["text"] = " " * len(row["text"])
+    elif row["conf"] < 10 and len(row["text"]) >= 3:
+        bbox = (
+            row["left"] - 5,
+            row["top"] - 3,
+            row["left"] + row["width"] + 5,
+            row["top"] + row["height"] + 3,
+        )
+        word_df = pd.DataFrame(
+            pytesseract.image_to_data(
+                img.crop(bbox),
+                config="-c preserve_interword_spaces=1x1 --psm 7 -l eng",
+                output_type=Output.DICT,
+            )
+        )
+        new_words = " ".join(
+            word_df.loc[word_df["conf"] > 10, "text"].tolist()
+        )
+        if new_words:
+            row["text"] = new_words
+        else:
+            row["text"] = "□" * len(row["text"])
+    elif row["conf"] < 10:
+        # if the confidence is under 10 and its just three characters - box it
+        row["text"] = "□"
+    elif (
+        row["conf"] < 20
+        and len(row["text"]) == 1
+        and not row["text"].isalnum()
+    ):
+        # Artifacts from scans often appear as lone symbols
+        # if conf is low and they are all alone drop them
+        row["text"] = " "
+
+    return row["text"] + " "
+
+
+def add_newlines(row: pd.Series, state: dict) -> dict:
```

Comment:
The code in here is pretty tough to make sense of unless you're enmeshed in it. Can you add some comments, please explaining how it works and what it's trying to do?